File: mantra_dev/mantra_dev/report/payment_entry_details/payment_entry_details.py
# Copyright (c) 2025, Foram Shah and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
	columns = get_column()
	data = get_data()
	return columns, data

def get_column():
	colunms = []
	colunms.append({"label": "Payment Entry Name", "fieldname": "name", "fieldtype": "Link", "options": "Payment Entry", "width": 200})
	colunms.append({"label": "Payment Type", "fieldname": "payment_type", "fieldtype": "Data","width": 75})
	colunms.append({"label": "Party", "fieldname": "party", "fieldtype": "Data","width": 100})
	colunms.append({"label": "Party Name", "fieldname": "party_name", "fieldtype": "Data","width": 150})
	colunms.append({"label": "Remarks", "fieldname": "remarks", "fieldtype": "Data","width": 150})
	colunms.append({"label": "Narattion", "fieldname": "narration", "fieldtype": "Data", "width": 150})
	colunms.append({"label": "Type", "fieldname": "type", "fieldtype": "Data", "width": 150})
	colunms.append({"label": "Project", "fieldname": "project", "fieldtype": "Data", "width": 150})
	colunms.append({"label": "Paid Amount", "fieldname": "paid_amount", "fieldtype": "Currency", "width": 150})
	colunms.append({"label": "Update Remarks", "fieldname": "update_remarks", "fieldtype": "Button", "width": 150})
	colunms.append({"label": "Get Details", "fieldname": "get_details", "fieldtype": "Button", "width": 150})
	return colunms

def get_data():
	data = f"""
		SELECT
			py.name as name,
			py.payment_type as payment_type,
			py.party as party,
			py.party_name as party_name,
			py.remarks as remarks,
			py.custom_type as type,
			py.docstatus,
			py.custom_project_type as project,
			py.paid_amount as paid_amount,
			(SELECT GROUP_CONCAT(
				CASE 
					WHEN per.reference_doctype = 'Purchase Invoice' 
						THEN CONCAT('Purchase Invoice (', per.reference_name, '): ', (SELECT pi.custom_narration FROM `tabPurchase Invoice` pi WHERE pi.name = per.reference_name))
					WHEN per.reference_doctype = 'Expense Claim' 
						THEN CONCAT('Expense Claim (', per.reference_name, '): ', (SELECT ec.remark FROM `tabExpense Claim` ec WHERE ec.name = per.reference_name))
					WHEN per.reference_doctype = 'Purchase Order' 
						THEN CONCAT('Purchase Order (', per.reference_name, '): ', (SELECT po.custom_purpose FROM `tabPurchase Order` po WHERE po.name = per.reference_name))
					WHEN per.reference_doctype = 'Sales Invoice' 
						THEN CONCAT('Sales Invoice (', per.reference_name, '): ', (SELECT si.remarks FROM `tabSales Invoice` si WHERE si.name = per.reference_name))
					WHEN per.reference_doctype = 'Sales Order' 
						THEN CONCAT('Sales Order (', per.reference_name, '): ', (SELECT so.custom_special_instructions_for_bill_and_dc FROM `tabSales Order` so WHERE so.name = per.reference_name))
					ELSE 
						NULL
				END SEPARATOR ',')
			FROM `tabPayment Entry Reference` per 
			WHERE per.parent = py.name) AS narration,
			CASE
				WHEN py.remarks IS NOT NULL AND py.remarks != '' THEN
					CONCAT(
							'<button class="btn btn-primary pt-0 pb-0 update_remarks" style="background-color: #28a745;color: white;"',
								'data-name="', py.name, '" ',
								'data-remarks="', IFNULL(py.remarks, ''), '"',
								'data-narration="', IFNULL((
									SELECT GROUP_CONCAT(
										CASE 
											WHEN per.reference_doctype = 'Purchase Invoice' 
												THEN CONCAT('Purchase Invoice (', per.reference_name, '): ', (SELECT pi.custom_narration FROM `tabPurchase Invoice` pi WHERE pi.name = per.reference_name))
											WHEN per.reference_doctype = 'Expense Claim' 
												THEN CONCAT('Expense Claim (', per.reference_name, '): ', (SELECT ec.remark FROM `tabExpense Claim` ec WHERE ec.name = per.reference_name))
											WHEN per.reference_doctype = 'Purchase Order' 
												THEN CONCAT('Purchase Order (', per.reference_name, '): ', (SELECT po.custom_purpose FROM `tabPurchase Order` po WHERE po.name = per.reference_name))
											WHEN per.reference_doctype = 'Sales Invoice' 
												THEN CONCAT('Sales Invoice (', per.reference_name, '): ', (SELECT si.remarks FROM `tabSales Invoice` si WHERE si.name = per.reference_name))
											WHEN per.reference_doctype = 'Sales Order' 
												THEN CONCAT('Sales Order (', per.reference_name, '): ', (SELECT so.custom_special_instructions_for_bill_and_dc FROM `tabSales Order` so WHERE so.name = per.reference_name))
											ELSE 
												NULL
										END SEPARATOR '<br>') 
									FROM `tabPayment Entry Reference` per 
									WHERE per.parent = py.name
								), ''), '" ',
								'data-type="', IFNULL(py.custom_type, ''), '" ',
								'data-project="', IFNULL(py.custom_project_type, ''), '" ',
							'>Update Remarks</button>'
						)
					ELSE
						NULL
				END AS update_remarks,
				CASE
					WHEN py.name IS NOT NULL AND py.name != '' THEN
						CONCAT(
							'<button class="btn btn-primary pt-0 pb-0 get_details" style="background-color: #007bff;color: white;" ',
								'data-name="', py.name, '" ',
							'>Get Details</button>'
						)
					ELSE NULL
            	END AS get_details
		FROM 
			`tabPayment Entry` as py
		WHERE
			py.docstatus = 1
	"""

	data = frappe.db.sql(data,as_dict=True)
	return data

@frappe.whitelist()
def update_remarks(payment_entry=None,remarks=None,type=None,project=None):
	#update remarks in Payment Entry 
	try:
		py_doc = frappe.get_doc("Payment Entry",payment_entry)
		if not py_doc:
			return {"status": "error", "message": "Payment Entry not found"}
		if remarks:
			frappe.db.set_value("Payment Entry", payment_entry, "remarks", remarks)
		if type:
			frappe.db.set_value("Payment Entry", payment_entry, "custom_type", type)
		if project:
			frappe.db.set_value("Payment Entry", payment_entry, "custom_project_type", project)
		frappe.db.commit()
		return {"status": "success", "message": "Payment Entry updated successfully"}
	except Exception as e:
		return {"status": "error", "message": str(e)}

File: mantra_dev/mantra_dev/report/_payment_entry_details/payment_entry_details.js
// Copyright (c) 2025, Foram Shah and contributors
// For license information, please see license.txt

frappe.query_reports["Payment Entry Details"] = {
	"filters": [

	],
	onload: function () {
		
		//Update Remarks Button Click Event 
		$(document).on("click", ".update_remarks", function () {
			var button = $(this);
			var PaymentEntry = button.data("name");
			var Remarks = button.data("remarks");
			var Narration = button.data("narration");
			var Type = button.data("type") || "0";
			var Project = button.data("project") || "0";
			
			let d = new frappe.ui.Dialog({
				title: 'Update Remarks',
				fields: [
					{
						label: 'Payment Entry',
						fieldname: 'payment_entry',
						fieldtype: 'Data',
						read_only: 1,
						default: PaymentEntry
					},
					{
						label: 'Project And Type',
						fieldtype: 'Section Break'
					},
					{
						label: 'Type',
						fieldname: 'type',
						fieldtype: 'Data',
						default: Type,
						read_only: 0
					},
					{
						fieldtype: 'Column Break'
					},
					{
						label: 'Project',
						fieldname: 'project',
						fieldtype: 'Data',
						default: Project,
						reqd: 0,
					},
					{
						fieldtype: 'Section Break'
					},
					{
						label: 'Remarks',
						fieldname: 'remarks',
						fieldtype: 'Small Text',
						reqd: 0,
						default: Remarks
					},
					{
						fieldtype: 'Section Break'
					},
					{
						label: 'Narration',
						fieldname: 'narration',
						fieldtype: 'Small Text',
						read_only:1,
						reqd: 0,
						default: Narration
					},
				],
				primary_action_label: 'Update',
				primary_action(values) {
					frappe.call({
						method: "mantra_dev.mantra_dev.report.payment_entry_details.payment_entry_details.update_remarks",
						args:{
							payment_entry: values.payment_entry,
							remarks: values.remarks,
							type: values.type,
							project: values.project
						},
						callback : function(r){
							if (r.message.status === "success") {
								frappe.msgprint(`Payment Entry <b>${values.payment_entry}</b> updated successfully!`)
								frappe.query_report.refresh();
							}else{
								frappe.msgprint("Error: " + response.message.message);
							}
						}
					});
					d.hide()
				}
			});
			d.show();
		});

		// Get Details Button Click Event 
		$(document).on("click", ".get_details", function () {
			var button = $(this);
			var paymentEntryId = button.data("name");

			if (!paymentEntryId) {
				frappe.msgprint(__('Payment Entry ID not found.'));
				return;
			}
			frappe.call({
				method: "mantra_dev.mantra_dev.page.payment_page.payment_page.get_payment_entry_reference_details",
				args: { payment_entry: paymentEntryId },
				callback: function(r) {
					if (r.message) {
						if (r.message.error) {
							frappe.msgprint(r.message.error);
							return;
						}
						let referenceDetails = r.message.reference_details || [];
						let customDetails = r.message.custom_details || {};
		
						let referenceTableRows = referenceDetails.map(item => `
							<tr>
								<td style="text-align:left;">${(item["Reference ID"] || "N/A").split(",").join("<br>")}</td>
								<td style="text-align:left;">${(item["Doctype"] || "N/A").split(",").join("<br>")}</td>
								<td style="text-align:left;">${(item["Approvers"] || "No Approvers").split(",").join("<br>")}</td>
								<td style="text-align:left;">${(item["Approver Names"] || "N/A").split(",").join("<br>")}</td>
	
							</tr>
						`).join("");
		
						let customDetailsTable = `
							
							<table class="custom-details-table">
								<thead>
									<tr>
										<th>Type</th>
										<th>Project Type</th>
										<th>Approved By</th>
										<th>Remarks</th>
									</tr>
								</thead>
								<tbody>
									<tr>
										<td style="text-align:left;">${customDetails.custom_type || "N/A"}</td>
										<td style="text-align:left;">${customDetails.custom_project_type || "N/A"}</td>
										<td style="text-align:left;">${customDetails.custom_approved_by || "N/A"}</td>
										<td style="text-align:left;">${customDetails.remarks || "N/A"}</td>
									</tr>
								</tbody>
							</table>
						`;
		
						let d = new frappe.ui.Dialog({
							title: __("Reference Details"),
							size: "extra-large",
							fields: [{
								fieldtype: "HTML",
								options: `
									<style>
										.reference-data, .custom-details-table {
											border-collapse: collapse;
											width: 100%;
											font-size: 14px;
											
										}
										.reference-data th, .reference-data td, 
										.custom-details-table th, .custom-details-table td {
											padding: 10px;
											border: 1px solid #ddd;
											text-align: center;
										}
										.custom-details-table thead{
											font-weight: #FFFFFF;;
											background-color: #007cc3;
										}
										.reference-data thead {
											font-weight: #FFFFFF;;
											background-color: #007cc3;    
										}
										.section-title {
											font-weight: bold;
											font-size: 16px;
											margin-top: 10px;
											margin-bottom: 5px;
										}
									</style>
		
									<div class="section-title"></div>
									<table class="reference-data">
										<thead>
											<tr>
												<th>Reference ID</th>
												<th>Doctype</th>
												<th>Approvers</th>
												<th>Approver Names</th>
											</tr>
										</thead>
										<tbody>
											${referenceTableRows || `<tr><td colspan="4">No Reference Details Found</td></tr>`}
										</tbody>
									</table>
		
									<div class="section-title"></div>
									${customDetailsTable}
								`
							}]
						});
						d.show();
					}
				}
			});
		})
	}
};
